src/semantics.rs

```diff
@@ -3,54 +3,53 @@ use crate::schema::Schema;
 use std::collections::HashMap;
 
 type ValidationResult = Result<(), String>;
+type ValidationHashMap = HashMap<String, usize>;
 
 pub trait Validate {
     fn validate(&self, schema: &Schema) -> ValidationResult;
 }
 
 pub trait FieldCounter {
-    fn add_to_counter(&self, map: &mut HashMap<String, usize>);
-    fn remove_from_counter(&self, map: &mut HashMap<String, usize>);
+    fn add_to_counter(&self, map: &mut ValidationHashMap);
+    fn remove_from_counter(&self, map: &mut ValidationHashMap);
 }
 
 impl FieldCounter for Expression {
-    fn add_to_counter(&self, map: &mut HashMap<String, usize>) {
+    fn add_to_counter(&self, map: &mut ValidationHashMap) {
+        use Expression::*;
+        use LogicalExpression::*;
```

Comment:
NIT: My suggestion would be not using ambiguous import but explicitly using `Expression::{And, Or, Not}` or something like that.